src/Features/LanguageServer/Protocol/Api/InProcLanguageServer.cs

```diff
@@ -0,0 +1,404 @@
+            Queue.RequestServerShutdown += RequestExecutionQueue_Errored;
+        }
+
+        /// <summary>
+        /// Handle the LSP initialize request by storing the client capabilities and responding with the server
+        /// capabilities.  The specification assures that the initialize request is sent only once.
+        /// </summary>
+        [JsonRpcMethod(Methods.InitializeName, UseSingleObjectParameterDeserialization = true)]
+        public Task<InitializeResult> InitializeAsync(InitializeParams initializeParams, CancellationToken cancellationToken)
+        {
+            try
+            {
+                Logger?.TraceStart("Initialize");
+
+                Contract.ThrowIfTrue(_clientCapabilities != null, $"{nameof(InitializeAsync)} called multiple times");
+                _clientCapabilities = initializeParams.Capabilities;
+                return Task.FromResult(new InitializeResult
+                {
+                    Capabilities = _serverCapabilities,
+                });
+            }
+            finally
+            {
+                Logger?.TraceStop("Initialize");
+            }
+        }
+
+        [JsonRpcMethod(Methods.InitializedName)]
+        public virtual Task InitializedAsync()
+        {
+            return Task.CompletedTask;
+        }
+
+        [JsonRpcMethod(Methods.ShutdownName)]
+        public Task ShutdownAsync(CancellationToken _)
+        {
+            try
+            {
+                Logger?.TraceStart("Shutdown");
+
+                ShutdownImpl();
+
+                return Task.CompletedTask;
+            }
+            finally
+            {
+                Logger?.TraceStop("Shutdown");
+            }
+        }
+
+        protected virtual void ShutdownImpl()
+        {
+            Contract.ThrowIfTrue(_shuttingDown, "Shutdown has already been called.");
+
+            _shuttingDown = true;
+
+            ShutdownRequestQueue();
+        }
+
+        [JsonRpcMethod(Methods.ExitName)]
+        public Task ExitAsync(CancellationToken _)
+        {
+            try
+            {
+                Logger?.TraceStart("Exit");
+
+                ExitImpl();
+
+                return Task.CompletedTask;
+            }
+            finally
+            {
+                Logger?.TraceStop("Exit");
+            }
+        }
+
+        private void ExitImpl()
+        {
+            try
+            {
+                ShutdownRequestQueue();
+                JsonRpc.Dispose();
+            }
+            catch (Exception e) when (FatalError.ReportAndCatch(e))
+            {
+                // Swallow exceptions thrown by disposing our JsonRpc object. Disconnected events can potentially throw their own exceptions so
+                // we purposefully ignore all of those exceptions in an effort to shutdown gracefully.
+            }
+        }
+
+        [JsonRpcMethod(Methods.TextDocumentDefinitionName, UseSingleObjectParameterDeserialization = true)]
```

Comment:
yep, these are the ones that roslyn uses (not exhaustive).  Ideally the client would respect server capabilities and not send us messages we don't have support for, so I don't think this needs to be exhaustive (it would just fail later when it tries to find the handlers for the request).  

Custom messages are out of scope for now I think.

Reply:
> how do we expose the handler infra

Can't external people already export handlers? It's the same IVT as is needed to access the factory interface, right?

> Ideally the client would respect server capabilities and not send us messages we don't have support for

Sure, but with this we're allowing external specification of those server capabilities, right?

Reply:
>Can't external people already export handlers? It's the same IVT as is needed to access the factory interface, right?

Yeah, but the IVT is just so we don't have to make the factory interface public until we're sold on its design.  The current plan doesn't include making the handler infra public (though it may at some point in the future).  So I would consider custom message support as out of scope for now.

>Sure, but with this we're allowing external specification of those server capabilities, right?
That is true, but adding every method here wouldn't solve it since there would be no handlers either (until we support custom messages, and then we need this list to be extensible)

Reply:
Okay, cool. Makes sense. I guess I was looking at the whole thing as "internal API" not "some bits of it will be public but some won't".